### ISSUE TITLE
Add CANNON Debug Renderer

### ISSUE THREAD

**Issue by notchris** [react-spring/cannon-es#27]
I think it would be helpful for people who are testing different collision body types to include the debug renderer.  Great port btw!
[Official example](http://schteppe.github.io/cannon.js/tools/threejs/example.html)
<img src="https://i.imgur.com/cWGryo1.png"/>

**Comment by notchris** [comment]
I updated the file to work as an import if this is helpful.
`import CannonDebugRenderer from './CannonDebugRenderer';`

```js
import * as THREE from 'three';
import * as CANNON from 'cannon-es';

/**
 * Adds Three.js primitives into the scene where all the Cannon bodies and shapes are.
 * @class CannonDebugRenderer
 * @param {THREE.Scene} scene
 * @param {CANNON.World} world
 * @param {object} [options]
 */
export default function CannonDebugRenderer(scene, world, options) {
    options = options || {};

    this.scene = scene;
    this.world = world;

    this._meshes = [];

    this._material = new THREE.MeshBasicMaterial({ color: 0x00ff00, wireframe: true });
    this._sphereGeometry = new THREE.SphereGeometry(1);
    this._boxGeometry = new THREE.BoxGeometry(1, 1, 1);
    this._planeGeometry = new THREE.PlaneGeometry( 10, 10, 10, 10 );
    this._cylinderGeometry = new THREE.CylinderGeometry( 1, 1, 10, 10 );
}

CannonDebugRenderer.prototype = {

    tmpVec0: new CANNON.Vec3(),
    tmpVec1: new CANNON.Vec3(),
    tmpVec2: new CANNON.Vec3(),
    tmpQuat0: new CANNON.Vec3(),

    update () {

        const bodies = this.world.bodies;
        const meshes = this._meshes;
        const shapeWorldPosition = this.tmpVec0;
        const shapeWorldQuaternion = this.tmpQuat0;

        let meshIndex = 0;

        for (let i = 0; i !== bodies.length; i += 1) {
            const body = bodies[i];

            for (let j = 0; j !== body.shapes.length; j += 1) {
                const shape = body.shapes[j];

                this._updateMesh(meshIndex, body, shape);

                const mesh = meshes[meshIndex];

                if (mesh) {

                    // Get world position
                    body.quaternion.vmult(body.shapeOffsets[j], shapeWorldPosition);
                    body.position.vadd(shapeWorldPosition, shapeWorldPosition);

                    // Get world quaternion
                    body.quaternion.mult(body.shapeOrientations[j], shapeWorldQuaternion);

                    // Copy to meshes
                    mesh.position.copy(shapeWorldPosition);
                    mesh.quaternion.copy(shapeWorldQuaternion);
                }

                meshIndex += 1;
            }
        }

        for (let i = meshIndex; i < meshes.length; i += 1) {
            const mesh = meshes[i];
            if (mesh) {
                this.scene.remove(mesh);
            }
        }

        meshes.length = meshIndex;
    },

    _updateMesh (index, body, shape){
        let mesh = this._meshes[index];
        if (!this._typeMatch(mesh, shape)) {
            if (mesh) {
                this.scene.remove(mesh);
            }
            mesh = this._meshes[index] = this._createMesh(shape);
        }
        this._scaleMesh(mesh, shape);
    },

    _typeMatch (mesh, shape){
        if (!mesh) {
            return false;
        }
        const geo = mesh.geometry;
        return (
            (geo instanceof THREE.SphereGeometry && shape instanceof CANNON.Sphere) ||
            (geo instanceof THREE.BoxGeometry && shape instanceof CANNON.Box) ||
            (geo instanceof THREE.PlaneGeometry && shape instanceof CANNON.Plane) ||
            (geo.id === shape.geometryId && shape instanceof CANNON.ConvexPolyhedron) ||
            (geo.id === shape.geometryId && shape instanceof CANNON.Trimesh) ||
            (geo.id === shape.geometryId && shape instanceof CANNON.Heightfield)
        );
    },

    _createMesh (shape){
        const material = this._material;
        let geometry, geo, v0, v1, v2, mesh;

        switch(shape.type){

        case CANNON.Shape.types.SPHERE:
            mesh = new THREE.Mesh(this._sphereGeometry, material);
            break;

        case CANNON.Shape.types.BOX:
            mesh = new THREE.Mesh(this._boxGeometry, material);
            break;

        case CANNON.Shape.types.PLANE:
            mesh = new THREE.Mesh(this._planeGeometry, material);
            break;

        case CANNON.Shape.types.CONVEXPOLYHEDRON:
            // Create mesh
            geo = new THREE.Geometry();

            // Add vertices
            for (let i = 0; i < shape.vertices.length; i += 1) {
                const v = shape.vertices[i];
                geo.vertices.push(new THREE.Vector3(v.x, v.y, v.z));
            }

            for (let i=0; i < shape.faces.length; i += 1) {
                const face = shape.faces[i];

                // add triangles
                const a = face[0];
                for (let j = 1; j < face.length - 1; j += 1) {
                    let b = face[j];
                    let c = face[j + 1];
                    geo.faces.push(new THREE.Face3(a, b, c));
                }
            }
            geo.computeBoundingSphere();
            geo.computeFaceNormals();

            mesh = new THREE.Mesh(geo, material);
            shape.geometryId = geo.id;
            break;

        case CANNON.Shape.types.TRIMESH:
            geometry = new THREE.Geometry();
            v0 = this.tmpVec0;
            v1 = this.tmpVec1;
            v2 = this.tmpVec2;
            for (let i = 0; i < shape.indices.length / 3; i += 1) {
                shape.getTriangleVertices(i, v0, v1, v2);
                geometry.vertices.push(
                    new THREE.Vector3(v0.x, v0.y, v0.z),
                    new THREE.Vector3(v1.x, v1.y, v1.z),
                    new THREE.Vector3(v2.x, v2.y, v2.z)
                );
                let j = geometry.vertices.length - 3;
                geometry.faces.push(new THREE.Face3(j, j+1, j+2));
            }
            geometry.computeBoundingSphere();
            geometry.computeFaceNormals();
            mesh = new THREE.Mesh(geometry, material);
            shape.geometryId = geometry.id;
            break;

        case CANNON.Shape.types.HEIGHTFIELD:
            geometry = new THREE.Geometry();

            v0 = this.tmpVec0;
            v1 = this.tmpVec1;
            v2 = this.tmpVec2;
            for (let xi = 0; xi < shape.data.length - 1; xi += 1) {
                for (let yi = 0; yi < shape.data[xi].length - 1; yi += 1) {
                    for (let k = 0; k < 2; k += 1) {
                        shape.getConvexTrianglePillar(xi, yi, k===0);
                        v0.copy(shape.pillarConvex.vertices[0]);
                        v1.copy(shape.pillarConvex.vertices[1]);
                        v2.copy(shape.pillarConvex.vertices[2]);
                        v0.vadd(shape.pillarOffset, v0);
                        v1.vadd(shape.pillarOffset, v1);
                        v2.vadd(shape.pillarOffset, v2);
                        geometry.vertices.push(
                            new THREE.Vector3(v0.x, v0.y, v0.z),
                            new THREE.Vector3(v1.x, v1.y, v1.z),
                            new THREE.Vector3(v2.x, v2.y, v2.z)
                        );
                        let i = geometry.vertices.length - 3;
                        geometry.faces.push(new THREE.Face3(i, i+1, i+2));
                    }
                }
            }
            geometry.computeBoundingSphere();
            geometry.computeFaceNormals();
            mesh = new THREE.Mesh(geometry, material);
            shape.geometryId = geometry.id;
            break;
        }

        if (mesh) {
            this.scene.add(mesh);
        }

        return mesh;
    },

    _scaleMesh (mesh, shape) {
        let radius;
        switch (shape.type) {

        case CANNON.Shape.types.SPHERE:
            radius = shape.radius;
            mesh.scale.set(radius, radius, radius);
            break;

        case CANNON.Shape.types.BOX:
            mesh.scale.copy(shape.halfExtents);
            mesh.scale.multiplyScalar(2);
            break;

        case CANNON.Shape.types.CONVEXPOLYHEDRON:
            mesh.scale.set(1,1,1);
            break;

        case CANNON.Shape.types.TRIMESH:
            mesh.scale.copy(shape.scale);
            break;

        case CANNON.Shape.types.HEIGHTFIELD:
            mesh.scale.set(1,1,1);
            break;

        }
    }
};
```

**Comment by marcofugaro** [comment]
Yeah sure, it makes sense. We'd have to find a way though, since it has Three.js as a dependency.

**Comment by codynova** [comment]
[This](https://github.com/react-spring/use-cannon/blob/feature/ts-wireframes/src/wireframes.ts) is the strongly typed debugger I adapted for `cannon-es` and `use-cannon`. The only reason we haven't added it yet is that the worker presents some unique architectural issues. We can release a `cannon-es-debugger` module to avoid bringing Three as a production runtime dependency, or some other solution - but it still won't be usable with `use-cannon` until we can find time to refactor the architecture.

**Comment by marcofugaro** [comment]
Solution 2 would be to add CannonDebugRenderer directly into the source and mark Three.js as an _optional peerDependency_.

However I think the `cannon-es-debugger` approach is probably cleaner.

**Comment by codynova** [comment]
I created [react-spring/cannon-es-debugger](https://github.com/react-spring/cannon-es-debugger) and released `cannon-es-debugger@0.0.1-alpha.2`. I also fixed some pretty serious memory consumption issues that appeared to be long standing (although it's 4:30am here and I didn't dig too deeply). The debugger has some callbacks that might be handy as well. I'm closing this issue since the debugger exists for `cannon-es` now, but we probably want to mention it in the readme somewhere. Feedback is welcome.
